src/ui/TreeView.js

```diff
@@ -28,46 +29,44 @@ class TreeView {
 
   update (json) {
     if (this.view) {
-      yo.update(this.view, this.render(json), {onBeforeElUpdated: (fromEl, toEl) => {
-        toEl.style.display = fromEl.style.display
-        toEl.className = fromEl.className
-        return true
-      }})
+      yo.update(this.view, this.render(json))
     }
   }
 
-  renderObject (item, parent, key, expand) {
+  renderObject (item, parent, key, expand, keyPath) {
     var data = this.extractData(item, parent, key)
     var children = (data.children || []).map((child, index) => {
-      return this.renderObject(child.value, data, child.key, expand)
+      return this.renderObject(child.value, data, child.key, expand, keyPath + '_' + child.key)
     })
-    return this.formatDataInternal(key, data, children, expand)
+    return this.formatData(key, data, children, expand, keyPath)
   }
 
   renderProperties (json, expand) {
     var children = Object.keys(json).map((innerkey) => {
-      return this.renderObject(json[innerkey], json, innerkey, expand)
+      return this.renderObject(json[innerkey], json, innerkey, expand, innerkey)
     })
     return yo`<ul style=${this.cssUl}>${children}</ul>`
   }
 
-  formatDataInternal (key, data, children, expand) {
-    var label = yo`<span style=${this.cssLabel}><label style=${ui.formatCss(style.caret)}></label><span style=${ui.formatCss(style.data)}>${this.formatData(key, data)}</span></span>`
+  formatData (key, data, children, expand, keyPath) {
+    var label = yo`<span style=${this.cssLabel}><label style=${ui.formatCss(style.caret)}></label><span style=${ui.formatCss(style.data)}>${this.formatSelf(key, data)}</span></span>`
     var renderedChildren = ''
     if (children.length) {
       renderedChildren = yo`<ul style=${this.cssUl}>${children}</ul>`
-      renderedChildren.style.display = expand ? 'block' : 'none'
-      label.firstElementChild.className = expand ? 'fa fa-caret-down' : 'fa fa-caret-right'
+      renderedChildren.style.display = this.keyPaths[keyPath] ? this.keyPaths[keyPath] : expand ? 'block' : 'none'
```

Comment:
Don't you have to declare `this.keyPaths` somewhere?

I think it is better to abstract away what the actual effect on the UI classes is and just store true/false inside keyPaths. Furthermore, what about using `nodeIsExpanded` as a name?